ZWaveLib/ZWaveLib.csproj

```diff
@@ -1,148 +1,35 @@
-    <RootNamespace>ZWaveLib</RootNamespace>
-    <AssemblyName>ZWaveLib</AssemblyName>
-    <FileAlignment>512</FileAlignment>
-    <ReleaseVersion>1.0</ReleaseVersion>
-    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <TargetFrameworks>netstandard2.0;net40;net461</TargetFrameworks>
+    <NetStandardImplicitPackageVersion>2.0.0</NetStandardImplicitPackageVersion>
+    <GenerateAssemblyConfigurationAttribute>false</GenerateAssemblyConfigurationAttribute>
+    <GenerateAssemblyCompanyAttribute>false</GenerateAssemblyCompanyAttribute>
+    <GenerateAssemblyProductAttribute>false</GenerateAssemblyProductAttribute>
+    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
+    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
+    <GenerateAssemblyTitleAttribute>false</GenerateAssemblyTitleAttribute>
   </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
-    <DebugSymbols>true</DebugSymbols>
-    <DebugType>full</DebugType>
-    <Optimize>false</Optimize>
-    <OutputPath>bin\Debug\</OutputPath>
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
-    <ErrorReport>prompt</ErrorReport>
-    <WarningLevel>4</WarningLevel>
-    <DocumentationFile>bin\Debug\ZWaveLib.xml</DocumentationFile>
-    <NoWarn>1591</NoWarn>
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
-    <DebugType>pdbonly</DebugType>
-    <Optimize>true</Optimize>
-    <OutputPath>bin\Release\</OutputPath>
-    <DefineConstants>TRACE</DefineConstants>
-    <ErrorReport>prompt</ErrorReport>
-    <WarningLevel>4</WarningLevel>
-  </PropertyGroup>
-  <ItemGroup>
-    <Reference Include="NLog, Version=4.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c">
-      <HintPath>..\packages\NLog.4.5.10\lib\net45\NLog.dll</HintPath>
-      <Private>True</Private>
-    </Reference>
-    <Reference Include="SerialPortLib, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null">
-      <HintPath>..\packages\SerialPortLib.1.0.18\lib\SerialPortLib.dll</HintPath>
-      <Private>True</Private>
-    </Reference>
-    <Reference Include="System" />
-    <Reference Include="Microsoft.CSharp" />
-    <Reference Include="System.Configuration" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Data" />
-    <Reference Include="System.IO.Compression" />
-    <Reference Include="System.Runtime.Serialization" />
-    <Reference Include="System.ServiceModel" />
-    <Reference Include="System.Transactions" />
-    <Reference Include="System.Xml" />
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.2.0" />
+    <PackageReference Include="NLog.Extensions.Logging" Version="1.6.1" />
```

Comment:
Ok, I absolutely agree with using `ILoggerFactory`, we can work on this. Right now I needed this `zwavelib` upgrade because I converted HomeGenie to a `netcoreapp3.1` and I wanted to make sure that was a possible thing.

Now I have a new `netcore` branch in HomeGenie (pushing it soon) 100% working but:
- I had to drop the Statistics web service handler along with SQLite
- There are no more `External` libraries (finally!), everything is managed via nuget packages
- I had to rework the `CSharpEngine` to make it compile again `csharp` program apps
- To make the SerialPortLib work in Ubuntu as netcoreapp3.1 it needs a `libnserial.so.1` to be compiled manually because it looks like this library is not part of debian system yet (see notes in README.md file)
- I had to fork all `RaspberrySharp` libs and add support for latest RPi models, because raspberrysharp looks like is not mantained since years
- I had to drop some features in `UPnP` library in order to make it work with netcore.

Statistics will be re-implemented with a different approach: it will not be implemented in HG core but rather as small programs apps (maybe using LiteDB) that will carry on the duty of collecting data from modules and elaborate this data real-time then storing it as ready to use static data-sets that will not require any query.

Also, the new homegenie will not run as `root` user anymore, but as a `homegenie` user that it will be added to the system during the installation (I will probably also integrate this in the traditional 1.3.x HomeGenie before switching to 1.4.x).

I am finally working on adding `ZigBee` support too and also on adding improved support for `Philips Hue` via `Q42.HueApi`, maybe those will be the first program apps using the `nuget packages` functionality.

I also plan to rewrite the new UI (maybe using Angular 2 + Angular Material, or maybe using just zuix, not sure yet) and that will be mantained as a separate project.

Reply:
This sounds really cool!
I have some questions about a couple of statements.

> * I had to drop the Statistics web service handler along with SQLite

You can take a Statistics module from my fork https://github.com/Bounz/HomeGenie-BE/tree/master/HomeGenie/Database which uses LiteDB. But to be honest there were problems with stability of LiteDB (but maybe they are solved now in v5.x)

> * There are no more `External` libraries (finally!), everything is managed via nuget packages

🎉 

> * To make the SerialPortLib work in Ubuntu as netcoreapp3.1 it needs a `libnserial.so.1` to be compiled manually because it looks like this library is not part of debian system yet (see notes in README.md file)

If you are going to target HG to netcoreapp3.1 you can consider using `System.IO.Ports` from .Net Core 3.0. I've used `SerialPortStream` only to make the library compatible with netstandard2.0. 

> Statistics will be re-implemented with a different approach: it will not be implemented in HG core but rather as small programs apps (maybe using LiteDB) that will carry on the duty of collecting data from modules and elaborate this data real-time then storing it as ready to use static data-sets that will not require any query.

What do you think about using a separate solution for handling metrics from HG? Like `prometheus` and `grafana`? If use docker distribution it would be easy to run all services using `docker-compose`.

> Also, the new homegenie will not run as `root` user anymore, but as a `homegenie` user that it will be added to the system during the installation (I will probably also integrate this in the traditional 1.3.x HomeGenie before switching to 1.4.x).

And again I will advertise docker distribution 🙂. It will allow to incapsulate everything into docker image and the one dependency needed on the host machine would be a `docker` service. It's also possible to create docker images targeting to different cpu architectures (x86, arm and so on). I've implemented multi-arch docker build on Drone build server but it could be done on Appveyor too. Example of multi-arch images: https://hub.docker.com/r/bounz/hgbe/tags.

> I would also like to improve the program editor and allow csharp programs to add arbitrary dependencies from `nuget packages` so all other non-core libraries such as `RaspberrySharp` will be added runtime.

👍 

> I also plan to rewrite the new UI (maybe using Angular 2 + Angular Material, or maybe using just zuix, not sure yet) and that will be mantained as a separate project.

It would be great, but how would you solve problem with widget editor when using Angular app? Would it be possible to create angular components in runtime?

Reply:
>If you are going to target HG to netcoreapp3.1 you can consider using System.IO.Ports from .Net Core 3.0. I've used SerialPortStream only to make the library compatible with netstandard2.0.

I think, if you are still interested in this project, maybe we can join our efforts and you can contribute merging your custom enhancement to HG to the official repo?

> What do you think about using a separate solution for handling metrics from HG? Like prometheus and grafana? If use docker distribution it would be easy to run all services using docker-compose.

this is also something that can be considered.

> And again I will advertise docker distribution slightly_smiling_face. 

no problems at all dockerizing HG

> It would be great, but how would you solve problem with widget editor when using Angular app? Would it be possible to create angular components in runtime?

No, it's not possible to create components at runtime, but it's possible to dynamically allocate components that are a already compiled with the app.
I've been thinking about that and there are 2 solutions:
 1) dropping support for the Widget editor
 2) using `zuix` for that
I am currently in favour of the solution 1. It would be possible anyway to add new widgets to the angular app via its source code.
I have almost completed the z-wave configuration components in the new angular app =)